### PR TITLE
[Feature] Gigachat 3.5 support

### PR DESCRIPTION
**Pull request by GungnirAP**
## Motivation

GigaChat-3.5-432B-A28B is a Mixture-of-Experts (MoE) language model with 432B total parameters and 28B active parameters. Built on a DeepSeek-V3-style backbone (MLA attention + DeepSeek MoE), it employs a hybrid attention architecture in which most layers use a Qwen3-Next Gated-Delta-Net (GDN) linear-attention block while a periodic subset retains full MLA attention. The architecture further incorporates gated RMSNorm with a low-rank gating bottleneck arranged as a four-norm sandwich (pre/post attention, pre/post MLP), gated attention applied before the output projection, and learnable MLA query/key scaling factors (alpha_q / alpha_kv). It utilizes 2-layer Multi-Token Prediction (MTP) for accelerated inference via speculative decoding.

## Modifications

Support for ai-sage/GigaChat3.5-432B-A28B in sglang.
- Registers GigaChat35ForCausalLM and the GigaChat35ForCausalLMNextN draft model, with GigaChat35Config (model_type: gigachat3_5) wired into the config registry and
the MLA attention arch resolved for both.
- Implements the hybrid decoder: DeepSeek-style MLA full attention on a subset of layers and Qwen3-Next Gated-Delta-Net (GDN) linear attention on the rest,
DeepSeek-style routed + shared experts, and pre_post sandwich gated RMS norms; loads the FP8 block-quantized checkpoint through the DeepSeek-V2 weight loader with
GigaChat weight-name remapping.
- Registers the model in the linear-attention model registry so the GDN attention backend and mamba radix cache are selected, and derives the Mamba2 cache shapes from
the checkpoint's linear-attention fields.
- Serves the model's 2 MTP heads through the multi-layer EAGLE worker with hidden-state chaining (each speculative step consumes the previous head's hidden states);
--speculative-algorithm EAGLE automatically enables multi-layer EAGLE, and shared-experts fusion is disabled, via the model-override registry.
- Commits accepted per-step mamba/GDN states into the persistent caches after multi-layer EAGLE verify, so rejected draft tokens do not advance the linear-attention
state.
- Adds the gigachat35 tool-call parser for the GCML tool-call format, covering both non-streaming and streaming parsing.
- Adds GSM8K e2e tests for the base and 2-step-MTP configurations and updates the docs (supported-models table, tool-call parser list).

## Accuracy Tests

8×H100, 200-example GSM8K, temp=1.0, top_p=0.95:

| Config | Accuracy | Latency | Output throughput | Accept len |
|---|---:|---:|---:|---:|
| Base (TP8/EP8) | 0.97 | 102.8 s | 332 tok/s | — |
| MTP-2 | 0.98 | 76.7 s | 447 tok/s | ~2.85 / 3 (≈0.93) |

## Speed Tests and Profiling

8×H100, TP8/EP8, FP8, input 1024 / output 512:

| Batch | Base decode | MTP-2 decode | Speedup | Acc length |
|---:|---:|---:|---:|---:|
| 1 | 70.9 tok/s | 155.9 tok/s | 2.20× | 2.84 |
| 16 | 684.8 tok/s | 1310.6 tok/s | 1.91× | 2.86 |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #35438831509](https://github.com/sgl-project/sglang/actions/runs/35438831509)
Latest PR Test (Extra): :x: [Run #35438831132](https://github.com/sgl-project/sglang/actions/runs/35438831132)
Latest PR Test (AMD ROCm 10): :x: [Run #35438831275](https://github.com/sgl-project/sglang/actions/runs/35438831275)

### CODE REVIEW

**Comment by ajpqs** [comment]
8×H100, TP8/EP8, FP8, input 1024 / output 512

batch | Base decode | MTP-2 decode | speedup | acc length
-- | -- | -- | -- | --
1 | 70.9 tok/s | 155.9 tok/s | 2.20x | 2.84
16 | 684.8 tok/s | 1310.6 tok/s | 1.91x | 2.86


8×H100, 200-example gsm8k, temp=1.0, top_p=0.95:
  | Accuracy | Latency | Output throughput | Accept len
-- | -- | -- | -- | --
Base (TP8/EP8) | 0.97 | 102.8 s | 332 tok/s | —
MTP-2 | 0.98 | 76.7 s | 447 tok/s | ~2.85 / 3 (≈0.93)

**Comment by GungnirAP** [comment]
@CatherineSue, @merrymercy, hello!
Could one of you please trigger CI with `/tag-and-rerun-ci`?
My `/rerun-failed-ci` command was accepted by the bot, but the PR still does not have the `run-ci` label, so the label-gated checks remain failed/skipped.

File: docs/docs/supported-models/generative_models.mdx [hunk 1]
```diff
@@ -303,5 +303,10 @@ in the GitHub search bar.
       <td style={{padding: "9px 12px", backgroundColor: "rgba(255,255,255,0.05)"}}><code>JetBrains/Mellum2-12B-A2.5B-Base</code>, <code>JetBrains/Mellum2-12B-A2.5B-Thinking</code></td>
```
Comment (mickqian) [inline docs/docs/supported-models/generative_models.mdx:303]: could we have a cookbook page for this model?

File: python/sglang/srt/models/gigachat35.py [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+            build_norm(config, config.hidden_size) if self._use_post else None
+        )
+        self.post_feedforward_layernorm = (
+            build_norm(config, config.hidden_size) if self._use_post else None
+        )
+
+        self._attn_prepare_layernorm = (
+            self.input_layernorm if self._use_pre else GigaChat35PassthroughNorm()
+        )
+        self._mlp_prepare_layernorm = GigaChat35MlpPrepNorm(
```
Comment (JustinTong0323) [inline python/sglang/srt/models/gigachat35.py]: Keep the preparation wrappers local to the communicator setup: registering the existing norms again under _attn_prepare_layernorm and _mlp_prepare_layernorm creates duplicate state_dict keys.

File: python/sglang/srt/function_call/gigachat35_detector.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            self._buffer = self.bot_token + rest
+            current_text = self._buffer
+            if leading:
+                return StreamingParseResult(normal_text=leading)
+
+        if not hasattr(self, "_tool_indices"):
+            self._tool_indices = self._get_tool_indices(tools)
+
+        _, _, after_open = current_text.partition(self.bot_token)
+        invokes_body, _, _ = after_open.partition(self.eot_token)
```
Comment (JustinTong0323) [inline python/sglang/srt/function_call/gigachat35_detector.py]: Return ordinary text after the GCML closing marker in both parsing modes; discarding the suffix here and in detect_and_parse silently removes it from the response.

File: python/sglang/srt/function_call/gigachat35_detector.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            self._buffer = ""
+            return StreamingParseResult(normal_text=current_text)
+
+        if not self._tool_region_started:
+            leading, _, rest = current_text.partition(self.bot_token)
+            self._tool_region_started = True
+            self._buffer = self.bot_token + rest
+            current_text = self._buffer
+            if leading:
+                return StreamingParseResult(normal_text=leading)
```
Comment (JustinTong0323) [inline python/sglang/srt/function_call/gigachat35_detector.py:109]: Drain pending input in finish(): a final chunk containing leading text followed by a complete GCML call returns only the text here and loses the call, while ordinary text ending in a partial opening marker is also lost at stream end.

File: test/registered/e2e/models/test_gigachat35_fp8.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+from sglang.test.run_eval import run_eval
+from sglang.test.test_utils import (
+    DEFAULT_TIMEOUT_FOR_SERVER_LAUNCH,
+    DEFAULT_URL_FOR_TEST,
+    CustomTestCase,
+    popen_launch_server,
+)
+
+register_cuda_ci(est_time=300, stage="extra-b", runner_config="8-gpu-h200")
+
```
Comment (JustinTong0323) [inline test/registered/e2e/models/test_gigachat35_fp8.py]: Drop this file. An 8xH200 CI job per model is more than our CI capacity can absorb.

File: test/registered/e2e/models/test_gigachat35_fp8_mtp.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+from sglang.test.run_eval import run_eval
+from sglang.test.test_utils import (
+    DEFAULT_TIMEOUT_FOR_SERVER_LAUNCH,
+    DEFAULT_URL_FOR_TEST,
+    CustomTestCase,
+    popen_launch_server,
+)
+
+register_cuda_ci(est_time=300, stage="extra-b", runner_config="8-gpu-h200")
+
```
Comment (JustinTong0323) [inline test/registered/e2e/models/test_gigachat35_fp8_mtp.py]: Drop this file as well; the MTP variant doubles the 8xH200 cost.

**Review by JustinTong0323 — CHANGES_REQUESTED** [review]
Fix GCML streaming finalization and preserve text after tool calls before merging; the norm module aliases also duplicate state_dict entries. Please also take the two 8-GPU e2e jobs out of CI — our CI capacity is limited and cannot absorb per-model jobs of this size.

File: python/sglang/srt/function_call/gigachat35_detector.py (hunk 3 above)
Comment (GungnirAP) [reply python/sglang/srt/function_call/gigachat35_detector.py]: Good catch, thanks! Fixed it in [b5dfce20bd](https://github.com/sgl-project/sglang/commit/b5dfce20bd0cc1256af7fbe1bc4e6a417c1b9581).

File: python/sglang/srt/function_call/gigachat35_detector.py (hunk 4 above)
Comment (GungnirAP) [reply python/sglang/srt/function_call/gigachat35_detector.py:109]: Also fixed in [b5dfce20bd](https://github.com/sgl-project/sglang/commit/b5dfce20bd0cc1256af7fbe1bc4e6a417c1b9581)

File: test/registered/e2e/models/test_gigachat35_fp8_mtp.py (hunk 6 above)
Comment (GungnirAP) [reply test/registered/e2e/models/test_gigachat35_fp8_mtp.py]: Got it, dropped both tests.

File: python/sglang/srt/models/gigachat35.py (hunk 2 above)
Comment (GungnirAP) [reply python/sglang/srt/models/gigachat35.py]: Thank you for the finding! Fixed it in [b87db3be43](https://github.com/sgl-project/sglang/commit/b87db3be43ba6448bfbf1b1fa60ffef5c5e33fc2).

File: python/sglang/srt/function_call/gigachat35_detector.py (hunk 3 above)
Comment (GungnirAP) [reply python/sglang/srt/function_call/gigachat35_detector.py]: We revised these three paths closely and think they don't apply to GigaChat 3.5 in practice, because the output shape is fixed by the chat template rather than free-form:
  - Text after the closing </｜GCML｜tool_calls> — the template emits the tool-call block as the terminal content of the assistant turn, so the model never produces prose after it. There is no trailing text to preserve.
  - A partial opening marker left at stream end — the GCML delimiters are single special tokens, not characters the model can "almost" emit as ordinary text; a dangling <｜GCML prefix isn't a reachable state here.

Also, the current behavior mirrors the deepseekv32_detector.py / deepseekv4_detector.py implementations this parser was inspired by — they handle the same edge cases identically (no finish() drain, trailing text after the closing marker not re-emitted).

File: python/sglang/srt/function_call/gigachat35_detector.py (hunk 4 above)
Comment (GungnirAP) [reply python/sglang/srt/function_call/gigachat35_detector.py:109]: Please, see the comment above.

**Review by JustinTong0323 — COMMENTED** [review]
Reviewer context for the two shared-component changes in this PR (attention_registry.py, multi_layer_eagle_worker_v2.py) — see inline notes. Everything else is standard model-registration boilerplate.

File: python/sglang/srt/layers/attention/attention_registry.py [hunk 7]
```diff
@@ -518,8 +518,7 @@ def attn_backend_wrapper(runner: "ModelRunner", full_attn_backend: "AttentionBac
         else:
             spec_result = get_linear_attn_config(runner.model_config.hf_config)
             if spec_result is not None:
-                spec, _ = spec_result
-                cfg = runner.model_config
+                spec, cfg = spec_result
```
Comment (JustinTong0323) [inline python/sglang/srt/layers/attention/attention_registry.py:525]: The hybrid branch below reads `cfg.full_attention_layer_ids`, which only exists on the config object returned by `get_linear_attn_config()` — not on `runner.model_config`, which is what the old code used. This registry has no registered models on `main`; GigaChat 3.5 is its first consumer, so no existing model is affected by the change.

File: python/sglang/srt/speculative/multi_layer_eagle_worker_v2.py [hunk 8]
```diff
@@ -265,7 +266,7 @@ def _init_boundary_kv_fix_state(self):
             return
         if isinstance(self.req_to_token_pool, HybridReqToTokenPool):
             conv_state = self.req_to_token_pool.mamba_pool.mamba_cache.conv
-            self.draft_extend_num_warmup_tokens = conv_state[0].shape[2]
+            self.draft_extend_num_warmup_tokens = min(conv_state[0].shape[2:])
```
Comment (JustinTong0323) [inline python/sglang/srt/speculative/multi_layer_eagle_worker_v2.py:272]: GigaChat 3.5's mamba conv state has a multi-dim trailing shape, so `shape[2]` alone is wrong here. For existing models with a single trailing dim, `min(shape[2:])` reduces to the old value — no behavior change.

File: python/sglang/srt/speculative/multi_layer_eagle_worker_v2.py [hunk 9]
```diff
@@ -1054,7 +1055,11 @@ def spec_v2_attn_backends(self) -> tuple:
         )
 
     def forward_batch_generation(
-        self, batch: ScheduleBatch, on_publish=None, grammar_barrier=None
+        self,
+        batch: ScheduleBatch,
+        on_publish=None,
+        grammar_barrier=None,
+        pp_proxy_tensors=None,
```
Comment (JustinTong0323) [inline python/sglang/srt/speculative/multi_layer_eagle_worker_v2.py:1065]: This hunk is a general latent bug, not GigaChat-specific: the scheduler's PP path passes `pp_proxy_tensors=` to every spec-v2 worker unconditionally (`scheduler.py:4514`, `:4527`), so any multi-layer-EAGLE model under PP crashes with TypeError on the first step — GigaChat3.5-432B is just the first model that requires it. The same parameter is missing in three other spec-v2 workers (`dspark_worker_v2` — #39262, `frozen_kv_mtp_worker_v2`, `uno_worker_v2`); umbrella issue #40155 tracks fixing all of them. Kept minimal here; can split into a standalone PR if preferred.

**Review by JustinTong0323 — APPROVED** [review]
